Append asset version to filename

clizzin opened the pull request:
Fixes #19. This works in development, but I have not tested in a staging/production environment yet because I don't have enough information to set up such an environment for myself. Figured I'd open this PR anyway so that someone else who has that ability can test this branch for me. Thanks!

clizzin committed:
[requirements.txt] Add Flask-Assets

clizzin committed:
Configure assets to include version hash in filename

Previously, if you edited an asset file, the server wouldn't deliver the
updated version, so you'd have to hard-refresh. This configures the
application to append a version hash to the asset filename so that
asset updates are requested and delivered as one would expect.

clizzin committed:
Add comment explaining how to enable debug mode for assets

jeredw commented:
can we also apply this to css? I've noticed some problems when editing the stylesheet too.

clizzin committed:
Configure CSS assets to include version hash in filename

clizzin commented:
@jeredw Great point. It is now done!

igor47 commented on spaceship/__init__.py:
```diff
@@ -1,5 +1,23 @@
 from flask import Flask
+from flask_assets import Environment, Bundle
```
to avoid polluting `__init__`, we should move all of this stuff into it's own file. for example, see `db.py`. we should definitely move asset initialization after we config the app via `app.config.from_object`.

igor47 commented on spaceship/__init__.py:
```diff
@@ -1,5 +1,23 @@
+
+# If you need the HTML to include each JS file via its own <script> tag,
+# uncomment this line:
+# assets.debug = True
```
is there a configuration option to the app that we can set in `config.py` to accomplish this? i wonder if we should default to `True` in dev but set it to `False` in prod?

igor47 reviewed (changes requested):
besides moving stuff around and configuring it, i think what's missing is build-time asset compilation. there should definitely be an entry in the `Dockerfile` to do that, probably after we `COPY .`. does flaask-assets provide some CLI to generate the bundles? i see that the underlying `webassets` command [does](https://webassets.readthedocs.io/en/latest/script.html), and we could maybe wrap it in an `invoke` task to make it more accessible

clizzin committed:
Refactor assets-related code into its own file

clizzin committed:
In production, use asset manifest file to locate pre-built asset bundles

We enable the manifest option so that the bundle filenames are recorded,
and we disable auto-build in production so that the server just uses the
manifest instead of attempting to build the bundles on the fly.

clizzin committed:
Build asset bundles and manifest in Docker image

Previously we relied on these files existing in the development
environment prior to building a deployment image (presumably
auto-generated while working on the project locally). Now we ensure that
the files are created as part of the build process.

igor47 reviewed (approved):
yay you did it!
